[Merged by Bors] - Require go 1.14.4

lrettig opened the pull request:
## Motivation
Closes #2050 (see that issue for background)
This resolves the CI issues I've been seeing

## Changes
- Switch travis from go 1.11.x to 1.14.4
- setup_env scripts check for go >= 1.14
- fix minor regexp bug in check-go-version script

lrettig committed:
Make install-protobuf antifragile

It should not fail if these dirs exist

lrettig committed:
Switch travis to go 1.13.x

See if this resolves CI issues

lrettig committed:
Update scrips to check for go 1.13+

lrettig committed:
Fix script regexp

lrettig commented:
@noamnelke how would you feel about bumping up the required min go version to 1.13?

noamnelke commented:
Oh... This is the first time I realize we declare we require go 1.13!! I personally have go 1.11.12 and never had issues. Why not just change the declaration in `go.mod` to `1.11`? OTOH, I don't really object to upgrading if you prefer.

noamnelke reviewed (approved):
I'm approving - up to you to decide if you want to merge

lrettig commented:
Thanks @noamnelke. go module support is complex and I've done my best to get up to speed here, but my understanding is that, while module support was only preliminary in go 1.11-1.12, in 1.13+ it's now considered production-ready. There are some [minor differences](https://github.com/golang/go/wiki/Modules) in how these different versions of go handle modules.

We definitely need to resolve the discrepancy I highlighted in #2050 (involving the [go directive](https://golang.org/doc/go1.12#modules) in go.mod). We could do it either way, by requiring go 1.13+ (as this PR does) or by "pinning" to go 1.11 somehow. I don't have a strong opinion either way, but other things being equal it seems to make sense to use the newer one.

@antonlerner @y0sher any opinions here? If not I'll go ahead and merge this

lrettig commented:
> Why not just change the declaration in `go.mod` to `1.11`?

I'm actually not totally sure this would work--the [instructions](https://golang.org/doc/go1.12#modules) on the go directive don't specify the behavior when running in a _newer_ version of go. I can test it.

lrettig committed:
Update README.md

Newer version of go

lrettig commented:
Didn't hear any objections, let's merge this

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Update go build in dockerfile

lrettig committed:
Update one more lingering version reference

lrettig commented:
bors merge

lrettig commented:
Getting an [unexpected error from bors build](https://travis-ci.org/github/spacemeshos/go-spacemesh/jobs/705959166). Need to investigate further:

```
 ---> Running in c812933195c8
./scripts/genproto.sh
using protoc from ./devtools/bin/protoc
Generating protobuf for api/pb
Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by ./devtools/bin/protoc)
Error relocating ./devtools/bin/protoc: __strftime_l: symbol not found
make: *** [Makefile:49: genproto] Error 127
The command '/bin/sh -c make build' returned a non-zero code: 2
Makefile:183: recipe for target 'dockerbuild-go' failed
make: *** [dockerbuild-go] Error 2
The command "make dockerpush" exited with 2.
```

Possible lead: https://github.com/bazelbuild/rules_closure/issues/228

lrettig committed:
Fix protoc docker build issues

The pre-built protoc binaries are not compatible with alpine linux.
Install the native alpine linux package instead, and modify the setup
scripts not to try to reinstall it.
Update protobuf include paths.

lrettig commented:
bors try

lrettig commented:
@noamnelke Would you mind taking another look at this one? I had to make a few more changes here to fix some Docker issues

noamnelke commented on scripts/genproto.sh:
```diff
@@ -18,7 +18,7 @@ errcho() {
 
 if [ ! -f $protoc ] ; then
-    errcho "Could not find protoc in $protoc. Trying to use protoc in PATH."
+    echo "Could not find protoc in $protoc. Trying to use protoc in PATH."
```
I think that leaving it red is better, here's why: if you've run `make install` you should have `protoc` in the propper place. This is probably bugging you (like me) because you're a legacy user. The way to fix it is to re-run `make install`. Why? Because using `protoc` from the `PATH` is error prone, you may not have the correct version, for example. Having this message stand out will make realizing this is the problem faster.

lrettig replied:
Will respond to your related point below

noamnelke commented on scripts/genproto.sh:
```diff
@@ -18,7 +18,7 @@ errcho() {
-    errcho "Could not find protoc in $protoc. Trying to use protoc in PATH."
+    echo "Could not find protoc in $protoc. Trying to use protoc in PATH."
     protoc=protoc # try loading from PATH
     if !(hash protoc 2>/dev/null) ; then
```
unrelated, but while you're in this area, you [should](https://github.com/koalaman/shellcheck/wiki/SC1035) add a space after the `!` here.

lrettig replied:
good catch, thank you, will fix

noamnelke commented on scripts/genproto.sh:
```diff
@@ -32,6 +32,7 @@ grpc_gateway_path=$(go list -m -f '{{.Dir}}' github.com/grpc-ecosystem/grpc-gate
-compile -I. -I$googleapis_path --go_out=plugins=grpc:. api/pb/api.proto
-compile -I. -I$googleapis_path --grpc-gateway_out=logtostderr=true:. api/pb/api.proto
-compile -I. -I$googleapis_path --swagger_out=logtostderr=true:. api/pb/api.proto
+echo protoc -I. -Idevtools/include -I$googleapis_path --go_out=plugins=grpc:. api/pb/api.proto
```
Wouldn't it be more useful to print:
```suggestion
echo $protoc -I. -Idevtools/include -I$googleapis_path --go_out=plugins=grpc:. api/pb/api.proto
```
(includes the full path that is actually run)

lrettig replied:
This was for debugging purposes, didn't mean to leave it in! Thanks for catching this

noamnelke commented on scripts/install-protobuf.sh:
```diff
@@ -16,17 +17,24 @@ echo "fetching ${protoc_url}"
 mkdir devtools/bin
 mkdir devtools/include
 
-echo "moving bin/protoc to ./devtools/bin/protoc"
-mv protoc3/bin/* devtools/bin/
+# Don't reinstall protoc if it's already installed
+if (hash protoc 2>/dev/null) ; then
+    echo "found systemwide protoc, not installing locally"
+else
+    echo "no systemwide protoc installation found"
+    echo "moving bin/protoc to ./devtools/bin/protoc"
+    mv protoc3/bin/* devtools/bin/
+fi
```
Not sure I agree with this change. We may want a version we control, even if it is installed globally.

lrettig replied:
I think it depends whether you feel that running a systemwide `protoc` is an error. I understand your point, that we may want to require a version we control, because a different version can result in different output that won't pass CI. However, this runs into two issues.

The first is FreeBSD, where at least the way it was set up in the PR we just merged, it relies on a systemwide installation: https://github.com/spacemeshos/go-spacemesh/pull/2058/files#diff-8269cac45755f62b44b6f8ab4426f294

The second is the upgraded Docker image in this PR. I'm actually not sure why this worked previously--maybe something changed in the more recent Docker images--but the binary protoc releases cannot be run on Alpine Linux because they're built using nonstandard musl libc. For this reason I modified the Dockerfile to instead install the systemwide package: https://github.com/spacemeshos/go-spacemesh/pull/2052/files#diff-3254677a7917c6c01f55212f86c57fbf

In any case, all of this will be going away in a matter of days when we merge the new API code, since we'll finally be able to remove all protobuf support, so I think there's no point in debating this!

noamnelke reviewed (changes requested):
One thing I'm considering, would like your opinion:

As long as we're updating the required Go version - shouldn't we update to the current latest (i.e. 1.14)? I'm not sure what's best here, we may not want to force devs to upgrade, but maybe we actually do?

lrettig commented:
> As long as we're updating the required Go version - shouldn't we update to the current latest (i.e. 1.14)? I'm not sure what's best here, we may not want to force devs to upgrade, but maybe we actually do?

Good question. For my part, I don't have a strong opinion about which version of Go we require or "pin" to, I just think we need to pick one to avoid the sort of issues I originally ran into here (building locally and building in CI producing different outputs). The choice of go 1.13 was sort of arbitrary. Glancing at the release notes ([1.13](https://golang.org/doc/go1.13), [1.14](https://golang.org/doc/go1.14)) it looks like 1.13 had more substantial changes, but 1.14 finalizes the design of the module system, which actually suggests that we should go with 1.14 since the issue that triggered this involves modules.

noamnelke commented:
Let's go with 1.14

lrettig committed:
Switch from go 1.13.12 to 1.14.4

lrettig commented:
bors try

lrettig committed:
Make linter happy

lrettig commented:
bors try

noamnelke commented on trie/database.go:
```diff
@@ -109,9 +109,11 @@ type rawShortNode struct {
-func (n rawShortNode) canUnload(uint16, uint16) bool { panic("this should never end up in a live trie") }
-func (n rawShortNode) cache() (hashNode, bool)       { panic("this should never end up in a live trie") }
-func (n rawShortNode) fstring(ind string) string     { panic("this should never end up in a live trie") }
+func (n rawShortNode) canUnload(uint16, uint16) bool {
+	panic("this should never end up in a live trie")
+}
+func (n rawShortNode) cache() (hashNode, bool)   { panic("this should never end up in a live trie") }
+func (n rawShortNode) fstring(ind string) string { panic("this should never end up in a live trie") }
```
Oopsie?

lrettig replied:
Not unrelated: https://travis-ci.org/github/spacemeshos/go-spacemesh/jobs/707748156. Unfortunately this is what happens when we change go versions, the linter expects different things.

noamnelke reviewed (approved):
I think you accidentally committed an unrelated change, otherwise good to go! :shipit:

lrettig commented:
bors merge

noamnelke commented on Makefile:
```diff
@@ -180,6 +180,8 @@ list-versions:
 
 
 dockerbuild-go:
+	# Must remove binaries or they will be copied into the container
+	rm -rf devtools/
```
This means we need to re-install after every docker build... better to add this folder to `.dockerignore`

Who approved this PR anyway?! 🤪

lrettig replied:
lol, will open a new issue now and will address this 👍
